### PR TITLE
Switch from QtSql to SQLiteCPP

### PR DESCRIPTION
**Pull request by cbielow**
## Description

Rewrite of OMSFileLoad and OMSFileStore to use the internal SQLiteCPP library instead of Qt's.

Fixes a few things in the implementation which Qt just ignored (e.g. nested transactions) or creating indices on non-existing tables.
Performance has improved (see below). Also correctness seems better since the benchmark below crashes for the old version on Windows. The new implementation runs fine.

Here are some benchmarks on debian/gcc. Input is a 60 MB OMS file (which I converted from a 446 MB idXML)
`old`
```
(base) ...w/openms_build $ ./bin/IDFileConverter -in ../../../tmp/OMS_bench/bla_qt.oms -out ../../../tmp/OMS_bench/bla_qt2.oms
Progress of 'Loading...':
-- done [took 5.39 s (CPU), 5.40 s (Wall)] --
Progress of 'Storing...':
-- done [took 5.16 s (CPU), 5.71 s (Wall)] --
IDFileConverter took 11.26 s (wall), 10.70 s (CPU), 0.22 s (system), 10.48 s (user); Peak Memory Usage: 188 MB.
```

`new`:
```
(base) ...w/openms_build $ ./bin/IDFileConverter -in ../../../tmp/OMS_bench/bla_qt.oms -out ../../../tmp/OMS_bench/bla_new2.oms
Progress of 'Loading...':
-- done [took 1.99 s (CPU), 2.00 s (Wall)] --
Progress of 'Storing...':
-- done [took 3.66 s (CPU), 3.93 s (Wall)] --
IDFileConverter took 6.02 s (wall), 5.75 s (CPU), 0.21 s (system), 5.54 s (user); Peak Memory Usage: 186 MB.
```

Thus, **read performance at factor 2.5x better, write performance at factor 1.4x better!** comparing the old QtSql implementation vs. [SQLiteCPP](https://github.com/SRombauts/SQLiteCpp). (@SRombauts may be happy to hear :) )
Note: I did not change the algorithm (apart from using std::move where necessitated by the new interface). This is just a conversion from one interface to another at this point.

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

**Comment by timosachsenberg** [comment]
@enetz can you run this through your benchmarking code you did when simplifying the schema?

**Comment by poshul** [comment]
Do we need to include the example code in third_parties?

File: src/openms/include/OpenMS/FORMAT/OMSFileLoad.h [hunk 1]
```diff
@@ -29,20 +29,19 @@
...
 
 #include <OpenMS/CONCEPT/ProgressLogger.h>
 #include <OpenMS/KERNEL/FeatureMap.h>
 #include <OpenMS/METADATA/ID/IdentificationData.h>
+#include <OpenMS/FORMAT/OMSFileStore.h>
 
-#include <QtCore/QJsonArray> // for JSON export
-#include <QString>
-#include <QStringList> // avoid compiler error on initialization using init. list
+#include <SQLiteCpp/Database.h>
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/OMSFileLoad.h:42]: could this be replaced by forward declarations?

File: src/openms/include/OpenMS/FORMAT/OMSFileLoad.h [hunk 2]
```diff
@@ -119,40 +118,40 @@ namespace OpenMS
...
 
-      void createView_(const QString& name, const QString& select);
+      void createView_(const String& name, const String& select);
 
       QJsonArray exportTableToJSON_(const QString& table, const QString& order_by);
 
-      // store name, not database connection itself (see https://stackoverflow.com/a/55200682):
-      QString db_name_;
+      /// The database connection (read)
+      SQLite::Database db_;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/OMSFileLoad.h:154]: would probably need a pointer type here

File: src/openms/include/OpenMS/FORMAT/OMSFileLoad.h (hunk 1 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/FORMAT/OMSFileLoad.h:42]: Was about to ask, then it could stay a private dependency.

**Comment by cbielow** [comment]
yes, replaced with PiMPL. Benchmark results did not change fortunately.

**Comment by timosachsenberg** [comment]
Is there a way to just fetch that code instead of including it?
I have no clue but some packages do something like
```
include(FetchContent)
FetchContent_Declare(
  xxx
  GIT_REPOSITORY xxxx
  GIT_TAG        vxxxx
)
FetchContent_MakeAvailable(xxxx)
```

**Comment by cbielow** [comment]
> Is there a way to just fetch that code instead of including it? I have no clue but some packages do something like
> 
> ```
> include(FetchContent)
> FetchContent_Declare(
>   xxx
>   GIT_REPOSITORY xxxx
>   GIT_TAG        vxxxx
> )
> FetchContent_MakeAvailable(xxxx)
> ```

Never tried actually.
If we move in this direction it would make sense to be consistent here IMHO, and do that with the other thirdparty libs as well.
But this is out of scope of this PR.

**Comment by cbielow** [comment]
> Do we need to include the example code in third_parties?

Does not hurt (apart from a few KB), and may be a nice reference for lookup in later sqlite implementation efforts.

**Comment by cbielow** [comment]
ready for merge. Tests pass (even though Windows CI seems down).
Any more comments/TODO's?

**Review by timosachsenberg — APPROVED** [review]
so libsqlite is now also included through thirdparty (not the contrib anymore)?

**Comment by cbielow** [comment]
> so libsqlite is now also included through thirdparty (not the contrib anymore)?

Yes. So we could remove it from contrib as well.

Other things which could be done (other PR):
 - remove our own sqliteadapter and replace with SQLiteCPP (mostly affects openswath code)

**Comment by timosachsenberg** [comment]
k I think this is good to go. I can remove slite from contrib as well as I got a PR open anyways.

**Comment by SRombauts** [comment]
Thanks @cbielow from pinging me, I am delighted to see your work has been merged, and as such SQLiteCpp

**Pull request merged**